util/src/rlp/rlpcompression.rs

```diff
@@ -0,0 +1,245 @@
+	pub fn new(rlps_to_swap: &[&'a [u8]], invalid_rlps: &[&'a [u8]]) -> Self {
+		if rlps_to_swap.len() > 0x7e {
+			panic!("Invalid usage, only 127 RLPs can be swappable.");
+		}
+		let mut invalid_to_valid = HashMap::new();
+		let mut valid_to_invalid = HashMap::new();
+		for (&rlp, &invalid) in rlps_to_swap.iter().zip(invalid_rlps.iter()) {
+			invalid_to_valid.insert(invalid, rlp);
+			valid_to_invalid.insert(rlp, invalid);
+		}
+		InvalidRlpSwapper {
+			invalid_to_valid: invalid_to_valid,
+			valid_to_invalid: valid_to_invalid
+		}
+	}
+	/// Get a valid RLP corresponding to an invalid one
+	fn get_valid(&self, invalid_rlp: &[u8]) -> Option<&[u8]> {
+		self.invalid_to_valid.get(invalid_rlp).map(|r| r.clone())
+	}
+	/// Get an invalid RLP corresponding to a valid one
+	fn get_invalid(&self, valid_rlp: &[u8]) -> Option<&[u8]> {
+		self.valid_to_invalid.get(valid_rlp).map(|r| r.clone())
+	}
+}
+
+/// Type of RLP indicating its origin database.
+pub enum RlpType {
+	/// RLP used in blocks database.
+	Blocks,
+	/// RLP used in snapshots.
+	Snapshot,
+}
+
+fn to_elastic(slice: &[u8]) -> ElasticArray1024<u8> {
+	let mut out = ElasticArray1024::new();
+	out.append_slice(slice);
+	out
+}
+
+fn map_rlp<F>(rlp: &UntrustedRlp, f: F) -> Option<ElasticArray1024<u8>> where
+	F: Fn(&UntrustedRlp) -> Option<ElasticArray1024<u8>> {
+	match rlp.iter()
+	.fold((false, RlpStream::new_list(rlp.item_count())),
+		|(is_some, mut acc), subrlp| {
+  		let new = f(&subrlp);
+  		if let Some(ref insert) = new {
+  			acc.append_raw(&insert[..], 1);
+  		} else {
+  			acc.append_raw(subrlp.as_raw(), 1);
+  		}
+  		(is_some || new.is_some(), acc)
+  	  }) {
+  	(true, s) => Some(s.drain()),
+  	_ => None,
+  }
+}
+
+/// Replace common RLPs with invalid shorter ones.
+fn simple_compress(rlp: &UntrustedRlp, swapper: &InvalidRlpSwapper) -> ElasticArray1024<u8> {
+	if rlp.is_data() {
+		to_elastic(swapper.get_invalid(rlp.as_raw()).unwrap_or(rlp.as_raw()))
+	} else {
+		map_rlp(rlp, |r| Some(simple_compress(r, swapper))).unwrap_or(to_elastic(rlp.as_raw()))
+	}
+}
+
+/// Recover valid RLP from a compressed form.
+fn simple_decompress(rlp: &UntrustedRlp, swapper: &InvalidRlpSwapper) -> ElasticArray1024<u8> {
+	if rlp.is_data() {
+		to_elastic(swapper.get_valid(rlp.as_raw()).unwrap_or(rlp.as_raw()))
+	} else {
+		map_rlp(rlp, |r| Some(simple_decompress(r, swapper))).unwrap_or(to_elastic(rlp.as_raw()))
+	}
+}
+
+/// Replace common RLPs with invalid shorter ones, None if no compression achieved.
+/// Tries to compress data insides.
+fn deep_compress(rlp: &UntrustedRlp, swapper: &InvalidRlpSwapper) -> Option<ElasticArray1024<u8>> {
+	let simple_swap = ||
+		swapper.get_invalid(rlp.as_raw()).map(|b| to_elastic(&b));
+	if rlp.is_data() {
+		// Try to treat the inside as RLP.
+		return match rlp.payload_info() {
+			// Shortest decompressed account is 70, so simply try to swap the value.
+			Ok(ref p) if p.value_len < 70 => simple_swap(),
+			_ => {
+				if let Ok(d) = rlp.data() {
```

Comment:
line `109` already proves that rlp is data. You can use `expect(...)` here or replace line 109 with:

`if let Ok(ref d)  = rlp.data() {`